### PR TITLE
Feature/farm 351 profit and loss repo

### PR DESCRIPTION
**Pull request by LeeMatthewHiggins**
[FARM-351 - Profit and loss repository]

#### 📲 What

Added the Bloc and repo for profit and loss section
Includes Mock
Missing firebase implementation of repo (will be in another PR)

#### 🤔 Why
		
This is so we can test the business logic of the Profit and loss section, as well as testing the UI.  
		
#### 🛠 How
		
A template Mock list repo was created to share the concept of adding objects to a list repository.
My Plot Mock was updated to use this template and so all repos that have this pattern will share this code. (User profile will too).  

Integrated the UI to be able to test the business logic in app.

#### 🕵️‍♂️ How to test

Go to Profit and Loss tab 

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

File: lib/model/bloc/transactions/ProfitLossListProvider.dart [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  @override
+  ProfitLossListViewModel snapshot() {
+    return _snapshot;
+  }
+
+  @override
+  ProfitLossListViewModel initial() {
+    if (_snapshot == null) {
+      _taglist= _getTagList([]);
+      _transactionsRepository.observeProfile(null).listen((transactions) {
```
Comment (LeeMatthewHiggins) [inline lib/model/bloc/transactions/ProfitLossListProvider.dart:50]: for now I add null here, I need to integrate with the profile repo when that's done.

File: lib/model/bloc/transactions/ProfitLossListProvider.dart [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    LoadingStatus status,
+    List<ProfitLossListItemViewModel> items = const [],
+  }) {
+    final transformer = TransactionToRecordTransactionViewModel(
+        _transactionsRepository, _taglist);
+    final costViewModel = transformer.costViewModel();
+    final saleViewModel = transformer.saleViewModel();
+    return ProfitLossListViewModel(
+      title: _title,
+      detailText: _Strings.currentcyTag,
```
Comment (LeeMatthewHiggins) [inline lib/model/bloc/transactions/ProfitLossListProvider.dart]: this should be localised

File: lib/model/model/TransactionAmount.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    final decimal = Decimal.parse(value);
+    return TransactionAmount._(forceNegative ? -decimal : decimal);
+  }
+
+  String toString({bool allowNegative = false}) {
+    //LH this is a transaction amount, which is never displayed as negative number.
+    if(allowNegative) {
+      return _decimal.toString();
+    }
+    return _decimal.isNegative ? (-_decimal).toString():_decimal.toString();
```
Comment (LeeMatthewHiggins) [inline lib/model/model/TransactionAmount.dart]: need to add format rule to make sure   .x = .x0. i.e always has double digits

File: lib/model/repositories/MockListRepository.dart [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import 'dart:async';
+
+import 'package:farmsmart_flutter/model/model/EntityCollectionInterface.dart';
+import 'package:farmsmart_flutter/model/repositories/BasicRepositoryInterface.dart';
+typedef String IdentifyEntity<T>(T value);
+
+class MockListRepository<T> implements BasicRepositoryInterface<T> {
```
Comment (LeeMatthewHiggins) [inline lib/model/repositories/MockListRepository.dart:7]: this same pattern will be present in the firebase version. so should be quick to add firebase repos.

File: lib/model/repositories/transaction/implementation/MockTransactionRepository.dart [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      }).reduce((a,b) {
+        return a + b;
+      });
+    });
+  }
+
+  @override
+  Future<TransactionAmount> thisWeekCosts() {
+    // TODO: implement thisWeekCosts
+    return null;
```
Comment (LeeMatthewHiggins) [inline lib/model/repositories/transaction/implementation/MockTransactionRepository.dart:48]: post MVP

File: lib/ui/common/modal_navigator.dart [hunk 6]
```diff
@@ -1,25 +1,27 @@
 import 'package:flutter/material.dart';
 
-class ModalNavigator extends StatelessWidget {
-  ModalNavigator({
+class NavigationScope extends StatelessWidget {
+  NavigationScope({
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/modal_navigator.dart:4]: I decided to rename as it didn't have anything to do with Modal.. it just wrapped the widget in a Navigator....

File: lib/ui/common/modal_navigator.dart [hunk 7]
```diff
@@ -1,25 +1,27 @@
...
       onWillPop: () async => !await navigatorKey.currentState.maybePop(),
       child: Scaffold(
-        body: Navigator(
-          key: navigatorKey,
-          onGenerateRoute: (RouteSettings settings) {
-            WidgetBuilder builder = (BuildContext _) => child;
-            return MaterialPageRoute(builder: builder, settings: settings);
-          },
-        ),
+        body: child
```
Comment (LeeMatthewHiggins) [inline lib/ui/common/modal_navigator.dart:24]: for now. I dropped the navigator, as the back is broken.

File: lib/ui/common/ActionSheet.dart [hunk 8]
```diff
@@ -188,7 +188,7 @@ class _ActionSheetState extends State<ActionSheet> {
         child: RoundedButton(
             viewModel: RoundedButtonViewModel(
                 title: viewModel.cancelButtonTitle,
-                onTap: () => Navigator.pop(context)),
+                onTap: () => Navigator.of(context).pop()),
```
Comment (davidfranquet) [inline lib/ui/common/ActionSheet.dart:191]: 👍

File: lib/ui/common/ActionSheet.dart [hunk 9]
```diff
@@ -210,7 +210,7 @@ class _ActionSheetState extends State<ActionSheet> {
 
   void dismissActionSheet() {
     clearSelection();
-    Navigator.pop(context);
+    Navigator.of(context).pop();
```
Comment (davidfranquet) [inline lib/ui/common/ActionSheet.dart:213]: 👍

File: lib/model/model/TransactionAmount.dart [hunk 10]
```diff
@@ -0,0 +1,33 @@
+// DART HAS NO DECIMAL NUMBER TYPE!!!!!!!!!
```
Comment (davidfranquet) [inline lib/model/model/TransactionAmount.dart:1]: :/

**Review by davidfranquet — DISMISSED** [review]
LGTM

File: lib/model/bloc/transactions/TransactionToRecordTransactionViewModel.dart [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    return saleViewModel(from: from, editable: false);
+  }
+
+  RecordTransactionViewModel costViewModel(
+      {TransactionEntity from, bool editable = true}) {
+    final amount = from?.amount?.toString() ?? "";
+    return RecordTransactionViewModel(
+        amount: amount,
+        actions: _actions(from: from),
+        buttonTitle: _Strings.recordCost,
```
Comment (albertmiro) [inline lib/model/bloc/transactions/TransactionToRecordTransactionViewModel.dart]: This should be localised?

File: lib/model/bloc/transactions/TransactionToRecordTransactionViewModel.dart [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        isEditable: editable);
+  }
+
+  RecordTransactionViewModel saleViewModel(
+      {TransactionEntity from, bool editable = true}) {
+    final amount = from?.amount?.toString() ?? "";
+    return RecordTransactionViewModel(
+        amount: amount,
+        actions: _actions(from: from),
+        buttonTitle: _Strings.recordSale,
```
Comment (albertmiro) [inline lib/model/bloc/transactions/TransactionToRecordTransactionViewModel.dart]: This should be localised?

**Comment by davidfranquet** [comment]
@LeeMatthewHiggins -
this PR (https://github.com/farmsmart/farmsmart-flutter/pull/76 ) will have conflicts with yours (ActionSheet) if you want I can wait you to merge and then I'll rebase and resolve the conflicts.

File: lib/model/model/TransactionAmount.dart [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+  bool isSale() {
+    return !isCost();
+  }
+
+  bool isCost() {
+    return _decimal.isNegative;
+  }
+
+  TransactionAmount operator  +( covariant TransactionAmount other) => TransactionAmount._(_decimal + other._decimal);
```
Comment (albertmiro) [inline lib/model/model/TransactionAmount.dart:35]: 👍

File: lib/model/bloc/transactions/TransactionToRecordTransactionViewModel.dart (hunk 12 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/transactions/TransactionToRecordTransactionViewModel.dart]: 👍

File: lib/model/model/TransactionAmount.dart (hunk 3 above)
Comment (LeeMatthewHiggins) [reply lib/model/model/TransactionAmount.dart]: done

**Review by albertmiro — APPROVED** [review]
LGTM 🚀

**Review by davidfranquet — APPROVED** [review]

**Pull request merged**